v3.0.0

romanblanco opened the pull request:
## Problem
PR to release changes for v3.0.0.

### This concerns this parts of the OpenAPI Specification:
* [v3.0.0](https://spec.openapis.org/oas/v3.0.0)
* [v3.1.0](https://spec.openapis.org/oas/v3.1.0)

### The changes I made are compatible with:
- [x] OAS3
- [x] OAS3.1

### Related Issues
- https://github.com/rswag/rswag/discussions/678

### Checklist
- [x] Added tests
- [x] Changelog updated
- [x] Added documentation to README.md
- [x] Added example of using the enhancement into [test-app](https://github.com/rswag/rswag/tree/master/test-app)

Commit:
drop using #send

`#send` is dangerous to call on objects passed into your library.

Instead, this will take a page from normal RSpec request specs and allow
developers to specify a "request_params" hash and a "request_headers" hash in let blocks
that this lib can then access.

I had to use request_ because example#headers already exists in request
specs as a shorthand for integration_session.headers.

I think this also cleans up some of the logic nicely.

I moved extracting the params & headers into initializer to avoid
duplicating effort.

This will obviously break peoples' pre-existing specs, and needs to be
part of a major version change.

Commit:
fix unpassed query params

kwstannard committed:
fix for falsey param values

romanblanco committed:
cleanup: Drop Rspec v2 support

romanblanco committed:
cleanup: Drop support for OpenAPI (Swagger) v2

Zarthus committed:
upgrade to ruby 3.0.6 and rails 7.0.4.3 in local dev

Zarthus committed:
upgrade workflow versions

Zarthus committed:
latest rails version for 7

hashrocketeer committed:
Set swagger formatter as default for task

Co-authored-by: Matt Polito <matt.polito@hashrocket.com>

Zarthus committed:
remove ruby < 2.x deprecation warnings

stevegeek committed:
Improve generation of OpenAPI 3 requestBody for parameters

stevegeek committed:
Add tests and fixes to parsing of request body

stevegeek committed:
Add support for changing parameter type: :file to `type: :string, format: :binary` for OAS3

stevegeek committed:
Add test for bearer HTTP auth scheme

stevegeek committed:
Note that dry run will need disabling to capture responses as examples

stevegeek committed:
Fix test-app auth configuration for auth test and update openapi.json

romanblanco committed:
fix: correct mistage after conflict resolving

romanblanco committed:
cleanup: remove condition for deprecated Rails version

romanblanco committed:
revert: removed option to use :getter

romanblanco committed:
Merge pull request #693 from romanblanco/rebased-396-2

Rebase of #396

kwstannard committed:
fix README for new request_* pattern

romanblanco committed:
Merge pull request #694 from CredSimple/v3

fix README for new request_* pattern

romanblanco committed:
Merge branch 'master' into v3

romanblanco committed:
fix: correct scheme path

romanblanco committed:
Merge branch 'master' into v3

romanblanco committed:
Merge branch 'master' into v3

romanblanco committed:
Merge branch 'master' into v3

romanblanco committed:
cleanup: Remove code for deprecation

romanblanco committed:
docs: update CHANGELOG.md

romanblanco committed:
docs: note EOL date for Ruby

romanblanco committed:
fix: rebranding Swagger to OpenAPI

romanblanco commented:
@jtannas @mattpolito @BookOfGreg, I've tried to [test](https://github.com/RedHatInsights/compliance-backend/compare/master...romanblanco:rswag-v3) the changes for v3.0.0 and they seem to work.

romanblanco committed:
remove: remove deprecation warnings mocks in tests

romanblanco committed:
cleanup: remove unnecessary RAILS_VERSION constant

romanblanco committed:
Merge branch 'master' into v3

romanblanco committed:
ci: allow misspelled words

philsturgeon commented:
I'll be happy to test this if you can shout when alpha version is tagged.

fynsta commented on rswag-specs/lib/rswag/specs/request_factory.rb:
```diff
@@ -7,145 +7,120 @@
-        # OAS 3: https://swagger.io/docs/specification/serialization/
-        if swagger_doc && doc_version(swagger_doc).start_with?('3') && param[:schema]
+        # NOTE: https://swagger.io/docs/specification/serialization/
+        if param[:schema]
```
I can't add a comment down there, but shouldn't the `else` case be removed? Since it relies on `type` and `collectionFormat` which just do not exist anymore in the new OpenAPI version.

fynsta replied:
I also saw that RSwag explicitly supports using `type: ...` instead of the "correct" version`schema: { type: ... }`.  (in [this commit](https://github.com/rswag/rswag/commit/70eb277e049c95410aafb77807e227e4c85faebe)).

This actually leads to some confusing behaviour because in the current RSwag version: the only place where using `type: ` does not work is when trying to implement a specific serialization strategy (because then there is no `schema` key of course).

So I would move the
```
type = type = param[:type] || param.dig(:schema, :type)
```
to the top of the file instead, remove this condition altogether and then use that type for serialization.

romanblanco replied:
> https://github.com/rswag/rswag/pull/714/files#r1439746426

It can be removed. Thanks for pointing it out!

romanblanco replied:
Reminder to handle invalid attributes:

https://github.com/rswag/rswag/commit/812d311426c323c76d32ee7fdb17931d7f98f565#r148058990

stevegeek committed:
Rename methods to improve readability

stevegeek committed:
Improve readability of conditionals

stevegeek committed:
Cleanup test contexts

stevegeek committed:
Drop support for OAS V2 `body` attribute

BrianHawley commented on rswag-api/lib/generators/rswag/api/install/templates/rswag_api.rb:
```diff
@@ -4,11 +4,11 @@
   # NOTE: If you're using rswag-specs to generate Swagger, you'll need to ensure
   # that it's configured to generate files in the same folder
-  c.openapi_root = Rails.root.to_s + '/swagger'
+  c.openapi_root = Rails.root.to_s + '/openapi'
```
`Rails.root` is a Pathname, so it's better to do this:
```ruby
c.openapi_root = Rails.root.join('openapi').to_s
```

BrianHawley commented on .github/workflows/ruby.yml:
```diff
@@ -11,18 +11,8 @@ jobs:
-        # Rails 7.0.1 requires >= 2.7.0, so excluding 2.6
-        - rails: 7.0.1
-          ruby: 2.6
+        ruby: [3.0, 3.1, 3.2] # 3.0 EOL is 2024-03-31
```
Use "3.0" so it doesn't get converted to 3, which at this point would translate to 3.3. Also, add 3.3.

BrianHawley replied:
At this point 3.0 was EOL way too recently, so it's probably a good idea to keep support for it.

stevegeek committed:
Merge branch 'master' into v3_update

stevegeek committed:
Gemfile tweak

stevegeek committed:
Readd lost changes from #429 from masamichhhhi/add-enum-description-for-#201

stevegeek committed:
Update test for rails upgrade in test and .ruby-version

stevegeek committed:
lnstall sqlite 1.4 on older Rails

stevegeek committed:
Update spec to use new request_headers

stevegeek committed:
Remove deprecated getter option from example/spec

stevegeek committed:
Remove empty file added by merge

stevegeek committed:
Update docs to reflect use of request_params and that RSwag seems to expect parameter names are strings

romanblanco committed:
Merge pull request #774 from stevegeek/v3_update

Update V3 branch

romanblanco committed:
Merge remote-tracking branch 'upstream/master' into merge-master-into-v3

romanblanco committed:
Delete deprecated option openapi_strict_schema_validation

romanblanco committed:
Merge pull request #784 from romanblanco/merge-master-into-v3

Merge master into v3

stevegeek committed:
Merge branch 'v3' into v3_requestbody_improvements

# Conflicts:
#	rswag-specs/lib/rswag/specs/openapi_formatter.rb

stevegeek committed:
Fix spec and update json

stevegeek committed:
Parameters using in: :body still should parse to requestBody

stevegeek committed:
Add another test

stevegeek committed:
Fixing handling of in: :body parameters and of required attributes

stevegeek committed:
Correctly handle enum

stevegeek committed:
Remove the old "v3" spec in favour of the current integration spec and openapi formatter spec

stevegeek committed:
We should parse parameters even if no media type specified for body/form data

stevegeek committed:
Fix to correctly parse type of header params

stevegeek committed:
Fix description on schema only when a body parameter

stevegeek committed:
Spell fix

romanblanco committed:
Merge pull request #723 from stevegeek/v3_requestbody_improvements

Improving generation of  OAS3.0 `requestBody` from `parameter`s

romanblanco committed:
Merge branch 'master' into update-v3

romanblanco committed:
cleanup: Remove v2 specific parameter serialization

romanblanco committed:
refactor: Improve pathname

romanblanco committed:
Merge pull request #793 from romanblanco/update-v3

Merge master into v3 branch

romanblanco committed:
fix: Remove attribute accidentally left during merge

romanblanco committed:
fix: Move type definition into schema object

https://spec.openapis.org/oas/v3.1.0.html#fixed-fields-9
https://spec.openapis.org/oas/v3.1.0.html#schema-object
https://github.com/rswag/rswag/pull/429

romanblanco committed:
Merge pull request #794 from romanblanco/fix-merge

fix(v3): fix issues caused during merge and failing unit tests

romanblanco committed:
feat: raise if unsupported field for Parameter object is used

romanblanco committed:
Merge pull request #795 from romanblanco/raise-on-unsupported-param-type

feat: raise if unsupported field for Parameter object is used

romanblanco committed:
Merge branch 'master' into merge-master-v3

romanblanco committed:
Merge pull request #811 from romanblanco/merge-master-v3

Merge Rails 8 support into v3

romanblanco committed:
Merge branch 'master' into rm-ruby-version-lock

romanblanco committed:
Merge pull request #814 from romanblanco/rm-ruby-version-lock

Merge latest changes into v3

jtannas commented on CHANGELOG.md:
```diff
@@ -4,6 +4,24 @@ All notable changes to this project will be documented in this file.
+
+- Drop support for OpenAPI v2 (Swagger) (https://github.com/rswag/rswag/pull/574)
+- Drop Rspec V2 support (https://github.com/rswag/rswag/pull/636)
+- upgrade to ruby 3.0.6 and rails 7.0.4.3 in local dev (https://github.com/rswag/rswag/pull/625)
```
Let's make a note about dropping official support for Ruby 2.7 in this release.
I'd like to start using endless methods to help tidy up.

P.S. Hash literal omission from Ruby 3.1 would also be nice but 3.0 isn't EOL until 2 months from now.

jtannas reviewed (approved):
I don't have much to say about the v3 changes so far since it's I'm just getting back into the project.
I'm eager to get them into `master` though since I'd like to build on them for a more extensive project revamp.

ie. [my notes](https://fish-locket-90a.notion.site/RSwag-Notes-18b2f8f8fc4d806ab67df592f8cc7b3f)

Hopefully that happens soon :crossed_fingers:

jtannas committed:
Update gem description to reflect dropped support for OAS2

romanblanco committed:
Merge pull request #822 from rswag/v3_update-description

v3/Update gem description to reflect dropped support for OAS2

romanblanco committed:
Merge branch 'master' into fix-ci

romanblanco committed:
Merge pull request #831 from romanblanco/fix-ci

Merge CI fix into v3 branch

romanblanco commented:
> I'll be happy to test this if you can shout when alpha version is tagged.

@philsturgeon, thank you for offering your help!

The pre-release is now out: https://github.com/rswag/rswag/releases/tag/3.0.0.pre

philsturgeon commented:
@romanblanco congratulations! Massive effort! 

Getting from 3.0 to 3.1 was not super simple, but hopefully getting to 3.2 will be no effort at all in comparison: https://github.com/OAI/OpenAPI-Specification/releases/tag/3.2.0
